ngraph/core/reference/include/ngraph/runtime/reference/binary_convolution.hpp

```diff
@@ -0,0 +1,203 @@
+                                                      in_range(rel_i_z, {0, input_size_z}));
+
+                                                int f_buf_idx =
+                                                    (f_z * filter_size_y * filter_size_x) +
+                                                    (f_y * filter_size_x) + f_x;
+                                                int i_buf_idx =
+                                                    (rel_i_z * input_size_y * input_size_x) +
+                                                    (rel_i_y * input_size_x) + rel_i_x;
+
+                                                T_IN in_val = padding
+                                                                  ? static_cast<T_IN>(pad_value)
+                                                                  : static_cast<T_IN>(
+                                                                        input_channel[i_buf_idx]);
+
+                                                T_F f_val = extract_bit(
+                                                    filter_channel[f_buf_idx / n_bits],
+                                                    (n_bits - 1) - (f_buf_idx % n_bits));
+
+                                                if (in_val == static_cast<T_IN>(f_val))
+                                                {
+                                                    sum += static_cast<T_IN>(1);
+                                                }
+                                            }
+                                        }
+                                    }
+                                    input_channel += input_channel_size;
+                                    filter_channel += filter_channel_size;
+                                }
+                                *out = (2 * sum - static_cast<T_IN>(filter_channel_size));
+                                ++out;
+                            }
+                        }
+                    }
+                }
+            }
+            template <typename T_IN, typename T_F>
+            void binary_convolution(const T_IN* in,
+                                    const T_F* f,
+                                    T_IN* out,
+                                    const Shape& in_shape,
+                                    const Shape& f_shape,
+                                    const Shape& out_shape,
+                                    const Strides& strides,
+                                    const Strides& dilations,
+                                    const CoordinateDiff& pads_begin,
+                                    const CoordinateDiff& pads_end,
+                                    const float pad_value)
+            {
+                // this implementation supports 1D, 2D and 3D convolutions
+                NGRAPH_CHECK(in_shape.size() >= 3 && in_shape.size() <= 5,
```

Comment:
Please extract all of these NGRAPH_CHECKs into "validate" helper function.

Reply:
Done